src/codegen/llvm/codegen_llvm_helper_visitor.cpp

```diff
@@ -474,16 +475,56 @@ void CodegenLLVMHelperVisitor::convert_local_statement(ast::StatementBlock& node
 
         /// remove local list statement now
         std::unordered_set<nmodl::ast::Statement*> to_delete({local_statement.get()});
-        node.erase_statement(to_delete);
+        /// local list statement is enclosed in statement block
+        const auto& parent_node = dynamic_cast<ast::StatementBlock*>(local_statement->get_parent());
+        parent_node->erase_statement(to_delete);
 
         /// create new codegen variable statement and insert at the beginning of the block
         auto type = new ast::CodegenVarType(FLOAT_TYPE);
-        auto statement = std::make_shared<ast::CodegenVarListStatement>(type, variables);
-        const auto& statements = node.get_statements();
-        node.insert_statement(statements.begin(), statement);
+        auto new_statement = std::make_shared<ast::CodegenVarListStatement>(type, variables);
+        const auto& statements = parent_node->get_statements();
+        parent_node->insert_statement(statements.begin(), new_statement);
+    }
+}
+
+/**
+ * \brief Visit StatementBlock and rename all LOCAL variables
+ * @param node AST node representing Statement block
+ *
+ * Statement block in remainder loop will have same LOCAL variables from
+ * main loop. In order to avoid conflict during lookup, rename each local
+ * variable by appending unique number. The number used as suffix is just
+ * a counter used for Statement block.
+ *
+ * \todo : Check in LLVM if it is really necessary to rename all local
```

Comment:
LLVM does not really have names, but temporaries (`%1`, `%2`, etc.). Our convention is that we keep the names of variables (at least named ones: e.g. `%x`), reason being:

- Readability of IR (it still is not very readable but at least we can clearly see where is `voltage`, `node_id`, etc.)
- Ability to use `ValueSymbolTable` for the function to query a value by name. This class also does renaming, so conflicting entry would have an increment in it.

The problem is that there is only a "single" scope in our IR, i.e. some `x` in for loop (branch to a basic block) is visible further down in the subsequent basic blocks. As far as I am aware, LLVM's `Value`s fully operate based on `ValueSymbolTable`, so there is no alternative here I suppose (or I somehow missed that completely). Moreover, we need custom renaming anyway since we create an epilogue loop which is scalar, but with the same variables. Even if we reuse the LLVM infrastructure, we cannot rename on the fly as we generate code - we haven't constructed all IR yet so it has to be done on NMODL side.

An alternative would be to have a pointer from AST node to the current scope's symbol table, and on the lookup we can traverse it up through the chain of parents. But I think that there is no infrastructure in NMODL that supports that at the moment? And even if it did exist, we would still have to take care of renaming, so I suppose it is just a question of architecture and API.

Reply:
> Moreover, we need custom renaming anyway since we create an epilogue loop which is scalar, but with the same variables. Even if we reuse the LLVM infrastructure, we cannot rename on the fly as we generate code - we haven't constructed all IR yet so it has to be done on NMODL side.

This make sense!

> An alternative would be to have a pointer from AST node to the current scope's symbol table, and on the lookup we can traverse it up through the chain of parents. But I think that there is no infrastructure in NMODL that supports that at the moment? And even if it did exist, we would still have to take care of renaming, so I suppose it is just a question of architecture and API.

NMODL does have infrastructure to traverse parent nodes and lookup in parent scope etc. But, I looked at what you mentioned about clang behaviour:

```cpp
 void foo() {
     int x;
     x = 12;
     for(int i=0; i < 10; i++)
     {
         int x;
         x = 13;
     }
 }
```

```asm
→ clang -c test.c -S -emit-llvm && cat test.ll
; Function Attrs: noinline nounwind optnone ssp uwtable
define void @foo() #0 {
  %1 = alloca i32, align 4
  %2 = alloca i32, align 4
  %3 = alloca i32, align 4
  store i32 12, i32* %1, align 4
  store i32 0, i32* %2, align 4
  br label %4

4:                                                ; preds = %8, %0
  %5 = load i32, i32* %2, align 4
  %6 = icmp slt i32 %5, 10
  br i1 %6, label %7, label %11

7:                                                ; preds = %4
  store i32 13, i32* %3, align 4
  br label %8

8:                                                ; preds = %7
  %9 = load i32, i32* %2, align 4
  %10 = add nsw i32 %9, 1
  store i32 %10, i32* %2, align 4
  br label %4
```
So I have now no hesitation to pull this in NMODL as a separate pass. It would be cleaner and easy to debug.